README


This archive contains the distribution Run-Parts,
version 0.08:

  Offers functionality of Debian's run-parts tool in Perl

This software is copyright (c) 2014 by Axel Beckert.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for Run-Parts

0.10    UNRELEASED

        Replace "#!/usr/bin/perl" in test suite example scripts with
        "#!/usr/bin/env perl" to fix test suite failures on
        DragonFlyBSD which has no more compatibility link at
        /usr/bin/perl since recently.

        Fix issues reported by default perlcritic. (Patch by Alex
        Muntada, thanks!)

        Add repology package list badges to README.markdown.

        Switch bug tracker from rt.cpan.org to GitHub as rt.cpan.org
        is going away.

        Fix RT#127691 + GH#1 by checking, too, if /bin/run-parts is
        Debian's implementation.

        Fix link to Kwalitee status in README.

        Enable Dist::Zilla::Plugin::MetaProvides::Package in dist.ini.

0.09    2015-08-18 02:23:22+0200

        Replace Modern::Perl with "use 5.010;" + strict and warnings.
        Drop perlcritic.rc.

        Add release test to verify 5.010 compatibility (for now).

        Polish README a little bit.

        Move release tests to xt/release and drop %ENV checking code.

        Require Test::Warnings >= 0.005 for "warnings".

        Replace individual xt/release/kwalitee.t with
        Dist::Zilla::Plugin::Test::Kwalitee

0.08    2014-09-11 16:57:00+02:00

        Fix dist.ini to exclude emacs style backup files (except in
        the test data directory) as well as the cover_db.

        Hack around RT#66485 to be able to use done_testing() together
        with Test::NoWarnings. Also fixes number of skipped tests in
        t/basic-functionality.t on non-debianoid platforms.

        Fix Author vs Release test in t/manifest.t's skip message.

        Only run t/changes.t if $RELEASE_TESTING is set.

        Update t/win-*.t skip messages to in line with the others.

        Add a Test::Kwalitee based test.

0.07    2014-09-10 00:33:03+02:00

        Add test that the change log conforms to CPAN::Changes::Spec.

        Convert dates in change log to conform with
        CPAN::Changes::Spec.

        Require CPAN::Changes >= 0.23 for running the test suite.

        Switch to Dist::Zilla (with ExtUtils::MakeMaker) instead of
        Module::Build. Remove convenience Makefile.

        Fix Perl::Critic warning "Don't modify $_ in list functions"

        Consider Modern::Perl as ok for
        Perl::Critic::Policy::TestingAndDebugging::RequireUseStrict

        Use Travis CI for smoke testing and Coveralls.io for coverage
        reporting.

        Fix passing backends as reference. Add test for that.

        Use "use_ok" instead of "use" in all tests.

        Drop unreachable else clause while listing the directory
        contents.

        Simplify condition in Run::Parts::Debian::run_parts_command.

        Further new tests added:
        + Verify behaviour upon unknown backend parameter.
        + Verify that chomped_lines and lines die if called like a method.
        + Alternative backend names (module=perl, run-parts=debian)

0.06    2013-07-22 20:47:16+02:00

        Bump Module::Build requirement to 0.4004 for test_requires.

        Remove all environment variables listed as relevant for taint
        mode in perlsec(1) from the affected tests.

0.05    2013-07-21 17:56:01+02:00

        Fix test skipping numbers in t/basic-functionality.t.

        Documentation deduplication.

0.04    2013-07-10 19:55:57+02:00

        New method concat.

        New helper module Run::Parts::Common.

        Run::Parts::Perl now works properly on DOS-ish platforms

        Some refactoring which pulls in new dependencies.

0.03    2013-07-05 21:01:59+02:00

        Run::Parts has been split up into frontend and backend
        modules.

        A pure Perl backend module has been added.

        Run::Parts now can either choose the appropriate backend
        itself or use a given backend.

0.02    2013-06-08 20:04:53+02:00

	Skip functionality tests if /bin/run-parts is not there or not
	executable.

	Add POD encoding to make Test::Pod happy

	State in Build.PL that the test suite needs Test::Differences and
	that Run::Parts itself needs Perl >= 5.10.

        Disable Makefile.PL compatibility.

0.01    2013-06-07 20:26:57+02:00

        Initial release. More or less a proof of concept.


Build.PL
use Modern::Perl;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Run::Parts',
    license             => 'Perl_5',
    dist_author         => q{Axel Beckert <abe@deuxchevaux.org>},
    dist_version_from   => 'lib/Run/Parts.pm',
    release_status      => 'stable',
    create_readme       => 1,
    meta_merge => {
        resources => {
            repository => 'https://github.com/xtaran/run-parts',
            bugtracker => "https://rt.cpan.org/Dist/Display.html?Name=Run-Parts",
        }
    },
    configure_requires => {
        'Module::Build' => '0.4004',
        'Modern::Perl' => 0,
    },
    test_requires => {
        'Test::More' => 0,
        'Test::Differences' => 0,
        'Test::NoWarnings' => 0,
        'Test::CPAN::Changes' => 0.23,
    },
    requires => {
        perl => '5.10.1',
        autodie => 0,
        'Taint::Util' => 0,
        'File::Slurp' => 9999.17,
        'Exporter::Easy' => 0,
        'Scalar::Util' => 0,
    },
    develop_requires => {
        'Pod::Readme' => 0,
    },
    add_to_cleanup     => [
        'Run-Parts-*',
        'README',
    ],
);

$builder->create_build_script();


lib/Run/Parts.pm
package Run::Parts;

# ABSTRACT: Offers functionality of Debian's run-parts tool in Perl

use strict;
use warnings;
use 5.010;

use File::Slurp 9999.17;
use Run::Parts::Common;

# VERSION generated by DZP::OurPkgVersion

=encoding utf8

=head1 SYNOPSIS

Run::Parts offers functionality of Debian's L<run-parts(8)> tool in
Perl.

run-parts runs all the executable files named within constraints
described in L<run-parts(8)> and L<Run::Parts::Perl>, found
in the given directory.  Other files and directories are silently
ignored.

Additionally it can just print the names of the all matching files
(not limited to executables, but ignores blacklisted files like
e.g. backup files), but don't actually run them.

This is useful when functionality or configuration is split over
multiple files in one directory. A typical convention is that the
directory name ends in ".d". Common examples for such split
configuration directories:

    /etc/cron.d/
    /etc/apt/apt.conf.d/
    /etc/apt/sources.list.d/,
    /etc/aptitude-robot/pkglist.d/
    /etc/logrotate.d/
    /etc/rsyslog.d/

Perhaps a little code snippet.

    use Run::Parts;

    my $rp  = Run::Parts->new('directory'); # chooses backend automatically
    my $rpp = Run::Parts->new('directory', 'perl'); # pure perl backend
    my $rpd = Run::Parts->new('directory', 'debian'); # uses /bin/run-parts

    my @file_list        = $rp->list;
    my @executables_list = $rpp->test;
    my $commands_output  = $rpd->run;
    ...

=begin readme

=head1 INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

=end readme


=head1 BACKENDS

L<Run::Parts> contains two backend implementations.
L<Run::Parts::Debian> actually uses /bin/run-parts and
L<Run::Parts::Perl> is a pure Perl implementation of a basic set of
L<run-parts(8)>' functionality.

L<Run::Parts::Debian> may or may not work with RedHat's simplified
shell-script based reimplementation of Debian's L<run-parts(8)>.

By default L<Run::Parts> uses L<Run::Parts::Debian> if /bin/run-parts
exists, L<Run::Parts::Perl> otherwise. But you can also choose any of
the backends explicitly.


=for readme stop

=head1 METHODS

=head2 new (Constructor)

Creates a new L<Run::Parts> object. Takes one parameter, the directory
on which run-parts should work.

=cut

sub new {
    my $self = {};
    bless($self, shift);
    $self->{dir} = shift;

    my $backend = shift;
    if (defined $backend) {
        if (ref $backend) {
            $self->{backend} = ref($backend)->new($self->{dir});
        } elsif ($backend eq 'debian' or $backend eq 'run-parts') {
            use Run::Parts::Debian;
            $self->{backend} = Run::Parts::Debian->new($self->{dir});
        } elsif ($backend eq 'perl' or $backend eq 'module') {
            use Run::Parts::Perl;
            $self->{backend} = Run::Parts::Perl->new($self->{dir});
        } else {
            warn "Unknown backend $backend in use";
            require $backend;
            # uncoverable statement
            $self->{backend} = $backend->new($self->{dir});
        }
    } else {
        # uncoverable branch false
        if (-x '/bin/run-parts') {
            $self->{backend} = Run::Parts::Debian->new($self->{dir});
        } else {
            # uncoverable statement
            $self->{backend} = Run::Parts::Perl->new($self->{dir});
        }
    }

    return $self;
}

=head2 run_parts_command

Returns the L<run-parts(8)> command to run with the given command
parameter.

=cut

sub run_parts_command {
    my $self = shift;
    return $self->{backend}->run_parts_command(@_);
}

=head2 list

Lists all relevant files in the given directory. Equivalent to
"run-parts --list".

=cut

sub list {
    my $self = shift;
    return $self->run_parts_command('list');
}

=head2 test

Lists all relevant executables in the given directory. Equivalent to
"run-parts --test".

=cut

sub test {
    my $self = shift;
    return $self->run_parts_command('test');
}

=head2 run

Runs all relevant executables in the given directory. Equivalent to
"run-parts".

=cut

sub run {
    my $self = shift;
    return $self->run_parts_command();
}

=head2 concat

Returns the concatenated contents of all relevant files in the given
directory. Equivalent to "cat `run-parts --list`".

=cut

sub concat {
    my $self = shift;
    return lines(map { read_file($_, { chomp => 1 }) } $self->list());
}

=for readme continue

=head1 SEE ALSO

L<run-parts(8)>, L<Run::Parts::Debian>, L<Run::Parts::Perl>

=head1 BUGS

Please report any bugs or feature requests to C<bug-run-parts at
rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Run-Parts>.  I will
be notified, and then you'll automatically be notified of progress on
your bug as I make changes.


=head1 CODE

You can find a git repository of Run::Parts' code at
L<https://github.com/xtaran/run-parts>.


=head1 SUPPORT

=begin readme

You can find documentation for this module with the perldoc command.

    perldoc Run::Parts


You can also look for information at:

=end readme

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Run-Parts>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Run-Parts>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Run-Parts>

=item * Search CPAN

L<http://search.cpan.org/dist/Run-Parts/>

=back
=cut

42; # End of Run::Parts


lib/Run/Parts/Perl.pm
package Run::Parts::Perl;

# ABSTRACT: Pure Perl implementation of Debian's run-parts tool

use strict;
use warnings;
use 5.010;

use autodie;
use Taint::Util;
use Run::Parts::Common;

# VERSION generated by DZP::OurPkgVersion

=encoding utf8

=head1 SYNOPSIS

Pure Perl reimplementation of basic functionality of Debian's
L<run-parts(8)> tool.

L<run-parts(8)> runs all the executable files named within constraints
described below, found in the given directory.  Other files and
directories are silently ignored.

Additionally it can just print the names of the all matching files
(not limited to executables, but ignores blacklisted files like
e.g. backup files), but don't actually run them.

This is useful when functionality or configuration is split over
multiple files in one directory.

This module is not thought to be used directly and its interface may
change. See L<Run::Parts> for a stable user interface.

=head1 FILE NAME CONSTRAINTS

On unix-ish operating systems, the file name (but not the path) must
match ^[-A-Za-z0-9_]+$, i.e. may not contain a dot.

On dos-ish operating systems, the file name without suffix must match
^[-A-Za-z0-9_]+$, i.e. may not contain a dot. The suffix may contain
alphanumeric characters and is not mandatory. The full regular
expression the file name including the suffix must match is
^[-A-Za-z0-9_]+(\.[A-Za-z0-9]+)?$.

Debian's L<run-parts(8)> tool also offers to use alternative regular
expressions as file name constraints. This is not yet implemented in
L<Run::Parts::Perl>.

=cut

# On DOS and Windows, run-parts' regular expressions are not really
# applicable. Allow an arbitrary alphanumerical suffix there.
my $win_suffix = dosish() ? qr/\.[a-z0-9]+/i : qr'';
my $file_re = qr/^[-A-Za-z0-9_]+($win_suffix)?$/;

=head1 METHODS

=head2 new (Constructor)

Creates a new L<Run::Parts> object. Takes one parameter, the directory
on which run-parts should work.

=cut

sub new {
    my $self = {};
    bless($self, shift);
    $self->{dir} = shift;

    return $self;
}

=head2 run_parts_command

Executes the given action with the given parameters

=cut

sub run_parts_command {
    my $self = shift;
    my $rp_cmd = shift // 'run';

    my @result = $self->$rp_cmd(@_);

    return lines(@result);
}

=head2 list

Lists all relevant files in the given directory. Equivalent to
"run-parts --list". Returns an array.

=cut

sub list {
    my $self = shift;
    my $dir = $self->{dir};

    opendir(my $dh, $dir);
    my @list = sort map {
        # $dir can neither be '' nor undef, hence no check necessary
        "$dir/$_";
    } grep {
        /$file_re/
    } readdir($dh);
}

=head2 test

Lists all relevant executables in the given directory. Equivalent to
"run-parts --tests". Returns an array.

=cut

sub test {
    my $self = shift;
    my $dir = $self->{dir};

    return grep { -x } $self->list($dir);
}

=head2 run

Executes all relevant executables in the given directory. Equivalent to
"run-parts --tests". Returns an array.

=cut

sub run {
    my $self = shift;
    my $dir = $self->{dir};

    return map {
        my $command = $_;
        untaint($command);
        # uncoverable branch true
        $command =~ s(/)(\\)g if dosish();
        my $output = `$command`;
        chomp($output);
        $output;
    } $self->test($dir);
}

=head1 INTERNAL FUNCTIONS

=head2 dosish

Returns true if ran on a dos-ish platform, i.e. MS-DOS, Windows or
OS/2.

=cut

sub dosish {
    return $^O =~ /^(dos|os2|MSWin32)$/;
}

=head1 SEE ALSO

L<Run::Parts>, L<run-parts(8)>


=head1 BUGS

Please report any bugs or feature requests to C<bug-run-parts at
rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Run-Parts>.  I will
be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=cut

qr/\.d$/; # End of Run::Parts::Perl


lib/Run/Parts/Common.pm
package Run::Parts::Common;

# ABSTRACT: Common helpers for Run::Parts and its backends

use strict;
use warnings;
use 5.010;

use Exporter::Easy ( EXPORT => [qw[lines chomped_lines]] );
use Scalar::Util qw(blessed);

# VERSION generated by DZP::OurPkgVersion

=encoding utf8

=head1 SYNOPSIS

Exports helper functions used by L<Run::Parts> as well as its backends.

=head1 EXPORTED FUNCTIONS

=head2 lines

Gets an array of strings as parameter.

In scalar context returns a string with all lines concatenated. In
array context it passes through the array.

=cut

sub lines {
    # Sanity check
    die "lines is no method" if blessed $_[0];

    return wantarray ? @_ : join("\n", @_)."\n";
}

=head2 chomped_lines

Gets an array of strings as parameter and calls chomp() on it.

In scalar context returns a string with all lines concatenated. In
array context it passes through the array.

=cut

sub chomped_lines {
    # Sanity check
    die "chomped_lines is no method" if blessed $_[0];

    chomp(@_);
    return lines(@_);
}



=head1 SEE ALSO

L<Run::Parts>


=head1 BUGS

Please report any bugs or feature requests to C<bug-run-parts at
rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Run-Parts>.  I will
be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=cut

23; # End of Run::Parts::Common


lib/Run/Parts/Debian.pm
package Run::Parts::Debian;

# ABSTRACT: Perl interface to Debian's run-parts tool

use strict;
use warnings;
use 5.010;

use Run::Parts::Common;

# VERSION generated by DZP::OurPkgVersion

=encoding utf8

=head1 SYNOPSIS

Perl interface to Debian's L<run-parts(8)> tool. L<run-parts(8)> runs
all the executable files named within constraints described below,
found in the given directory.  Other files and directories are
silently ignored.

Additionally it can just print the names of the all matching files
(not limited to executables, but ignores blacklisted files like
e.g. backup files), but don't actually run them.

This is useful when functionality or configuration is split over
multiple files in one directory.

This module is not thought to be used directly and its interface may
change. See L<Run::Parts> for a stable user interface.

=head1 METHODS

=head2 new (Constructor)

Creates a new L<Run::Parts> object. Takes one parameter, the directory on
which run-parts should work.

=cut

sub new {
    my $self = {};
    bless($self, shift);
    $self->{dir} = shift;

    return $self;
}

=head2 run_parts_command

Returns the L<run-parts(8)> command to run with the given command
parameter

=cut

sub run_parts_command {
    my $self = shift;
    my $rp_cmd = shift;

    my $command =
        "/bin/run-parts " .
        ($rp_cmd ? "'--$rp_cmd'" : '') .
        " '".$self->{dir}."'";

    return chomped_lines(`$command`);
}

=head1 SEE ALSO

L<Run::Parts>, L<run-parts(8)>


=head1 BUGS

Please report any bugs or feature requests to C<bug-run-parts at
rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Run-Parts>.  I will
be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=cut

q<debian/rules>; # End of Run::Parts::Debian


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;



use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Offers functionality of Debian's run-parts tool in Perl",
  "AUTHOR" => "Axel Beckert <abe\@deuxchevaux.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Run-Parts",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Run::Parts",
  "PREREQ_PM" => {
    "Exporter::Easy" => 0,
    "File::Slurp" => "9999.17",
    "Modern::Perl" => 0,
    "Scalar::Util" => 0,
    "Taint::Util" => 0,
    "autodie" => 0
  },
  "TEST_REQUIRES" => {
    "Test::Differences" => 0,
    "Test::More" => 0,
    "Test::NoWarnings" => 0,
    "Test::Trap" => 0,
    "Test::Warnings" => 0
  },
  "VERSION" => "0.08",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter::Easy" => 0,
  "File::Slurp" => "9999.17",
  "Modern::Perl" => 0,
  "Scalar::Util" => 0,
  "Taint::Util" => 0,
  "Test::Differences" => 0,
  "Test::More" => 0,
  "Test::NoWarnings" => 0,
  "Test::Trap" => 0,
  "Test::Warnings" => 0,
  "autodie" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



